CheckRuntime(): fix panic and simplify logic

thaJeztah opened the pull request:
### CheckRuntime(): fix panic

There was an off-by-one error in the check, resulting in a panic if "expected" was longer than "current";

    CheckRuntime("a.b.c", "a.b.c.d")
    panic: runtime error: index out of range [3] with length 3

### CheckRuntime(): simplify logic

The intent of this function was to check if a given runtime has
the expected prefix. This patch updates the function to;

- disallow empty strings (either "current" or "expected")
- check for prefix or full match, making the code slightly more
  readable on intent.
- adds some more descriptive tests

thaJeztah commented:
oh, interesting; looks like there's another (non-exported) copy of this function, with the same issue. Perhaps it should be exported (or these should be moved somewhere else); https://github.com/containerd/containerd/blob/66994034035a0745d7ae3893db947b76a6ca2fab/services/tasks/local.go#L820-L835

fuweid commented:
> Nah why.
> 
> How about merging checkRuntime in this PR? This PR is still small enough.

https://github.com/containerd/containerd/pull/3004#discussion_r255786508 😂

fuweid commented on client_test.go:
```diff
@@ -0,0 +1,59 @@
+		{current: "io.containerd.runc.v1", prefix: "io.containerd", out: true},
+		{current: "io.containerd", prefix: "io.containerd.runc", out: false},
+		{current: "io.containerd", prefix: "io.containerd.runc.v1", out: false},
+		{current: "io.containerdruncv1", prefix: "io.containerd", out: false},
```
What if the `current: "containerdruncv1", prefix: "containerdruncv1"`? It might be edge case
Do we need to consider that?

ruiwen-zhao replied:
> current: "containerdruncv1", prefix: "containerdruncv1" 

This is covered by `current == prefix`, and should return true. Isnt this what we want?

samuelkarp replied:
I think we're missing a test case where the inputs are equal.

ruiwen-zhao commented:
/lgtm

thaJeztah commented:
> Let me have a look (I recall I wanted to do that, but wanted to look first if it was duplicated for a reason, and would cause issues with imports pulling int "too much")

So, yes, unifying is somewhat more involved; so I just updated the copy in `services/tasks`, but added a comment in the godoc to mention that it's a copy.

Perhaps we should consider moving the `Task`-related types and functions (`/task.go`, `/task_opts.go` etc.)  to a `task` package; containerd is quite "heavy" on what's in the root package, so having smaller / more focussed packages may not be a bad thing, but I'd have to check what that would look like.

Either ways; updated this PR; 

- added an extra test-case to address https://github.com/containerd/containerd/pull/7628#discussion_r1041629146
- updated both implementations to be the same

PTAL 🤗

mxpv commented:
We've removed the old runtimes in https://github.com/containerd/containerd/pull/8262 including `CheckRuntime` function. Should we close this PR?

thaJeztah commented:
Forgot about this one; may have another look this week
